clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -2276,17 +2279,26 @@ template <typename RecordType, typename ReaderType>
 void
 scheduler_tmpl_t<RecordType, ReaderType>::mark_input_eof(input_info_t &input)
 {
+    if (input.at_eof)
+        return;
     input.at_eof = true;
     assert(live_input_count_.load(std::memory_order_acquire) > 0);
     live_input_count_.fetch_add(-1, std::memory_order_release);
+    VPRINT(this, 2, "input %d at eof; %d live inputs left\n", input.index,
+           live_input_count_.load(std::memory_order_acquire));
 }
 
 template <typename RecordType, typename ReaderType>
 typename scheduler_tmpl_t<RecordType, ReaderType>::stream_status_t
 scheduler_tmpl_t<RecordType, ReaderType>::eof_or_idle(output_ordinal_t output)
 {
     if (options_.mapping == MAP_TO_CONSISTENT_OUTPUT ||
-        live_input_count_.load(std::memory_order_acquire) == 0) {
+        live_input_count_.load(std::memory_order_acquire) == 0 ||
+        // While a full schedule recorded should have each input hit either its
+        // EOF or ROI end, we have a fallback to avoid hangs for possible recorded
+        // schedules that end an input early deliberately without an ROI.
+        (options_.mapping == MAP_AS_PREVIOUSLY &&
+         live_replay_output_count_.load(std::memory_order_acquire) == 0)) {
         return sched_type_t::STATUS_EOF;
```

Comment:
Also assert for MAP_AS_PREVIOUSLY, we return STATUS_EOF iff at_eof is set.

Reply:
Maybe worth asserting that at_eof is set if we're returning STATUS_EOF?

Reply:
Done.